pkg/kafka/ingester/consumer.go

```diff
@@ -0,0 +1,158 @@
+package ingester
+
+import (
+	"bytes"
+	"context"
+	"crypto/rand"
+	"fmt"
+	"io"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/oklog/ulid"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"google.golang.org/grpc"
+
+	"github.com/grafana/loki/v3/pkg/ingester-rf1/metastore/metastorepb"
+	"github.com/grafana/loki/v3/pkg/kafka"
+	"github.com/grafana/loki/v3/pkg/logproto"
+	"github.com/grafana/loki/v3/pkg/storage/wal"
+)
+
+// ObjectStorage defines an interface for object storage operations
+type ObjectStorage interface {
+	PutObject(ctx context.Context, objectKey string, object io.Reader) error
+}
+
+// MetadataStore defines an interface for metadata storage operations
+type MetadataStore interface {
+	AddBlock(ctx context.Context, in *metastorepb.AddBlockRequest, opts ...grpc.CallOption) (*metastorepb.AddBlockResponse, error)
+}
+
+// consumer represents a Kafka consumer that processes and stores log entries
+type consumer struct {
+	metastoreClient MetadataStore
+	storage         ObjectStorage
+	writer          *wal.SegmentWriter
+
+	buf     *bytes.Buffer
+	decoder *kafka.Decoder
+	toStore []*logproto.Entry
+
+	metrics *consumerMetrics
+	logger  log.Logger
+}
+
+// newConsumer creates and initializes a new consumer instance
+func newConsumer(
+	metastoreClient MetadataStore,
+	storage ObjectStorage,
+	logger log.Logger,
+	reg prometheus.Registerer,
+) (*consumer, error) {
+	writer, err := wal.NewWalSegmentWriter()
+	if err != nil {
+		return nil, err
+	}
+	decoder, err := kafka.NewDecoder()
+	if err != nil {
+		return nil, err
+	}
+	return &consumer{
+		logger:          logger,
+		metastoreClient: metastoreClient,
+		storage:         storage,
+		writer:          writer,
+		metrics:         newConsumerMetrics(reg),
+		buf:             bytes.NewBuffer(make([]byte, 0, 10<<20)), // 10 MB
+		decoder:         decoder,
+	}, nil
+}
+
+// Consume processes a batch of Kafka records, decoding and storing them
+func (c *consumer) Consume(_ context.Context, _ int32, records []record) error {
+	for _, record := range records {
+		stream, labels, err := c.decoder.Decode(record.content)
+		if err != nil {
+			return fmt.Errorf("failed to decode record: %w", err)
+		}
+		if len(stream.Entries) == 0 {
+			continue
+		}
+		if len(c.toStore) == 0 {
+			c.toStore = make([]*logproto.Entry, 0, len(stream.Entries))
+		}
+		c.toStore = c.toStore[:0]
+		for _, entry := range stream.Entries {
+			c.toStore = append(c.toStore, &logproto.Entry{
+				Timestamp:          entry.Timestamp,
+				Line:               entry.Line,
+				StructuredMetadata: entry.StructuredMetadata,
+				Parsed:             entry.Parsed,
+			})
+		}
+		c.writer.Append(record.tenantID, stream.Labels, labels, c.toStore, time.Now())
+	}
+	return nil
+}
+
+// Flush writes the accumulated data to storage and updates the metadata store
+func (c *consumer) Flush(ctx context.Context) error {
```

Comment:
We should add a check for input size so we don't write empty blocks.

```diff
 // Flush writes the accumulated data to storage and updates the metadata store
 func (c *consumer) Flush(ctx context.Context) error {
+       if c.writer.InputSize() == 0 {
+               return nil
+       }
+
        start := time.Now()
        c.metrics.flushesTotal.Add(1)
        defer func() { c.metrics.flushDuration.Observe(time.Since(start).Seconds()) }()
```